wdl-lint/src/rules/nonmatching_output.rs

```diff
@@ -0,0 +1,249 @@
+        "Ensures that each output field is documented in the meta section under `meta.outputs`."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "The meta section should have an `outputs` key and keys with descriptions for each output of \
+         the task/workflow. These must match exactly. i.e. for each named output of a task or \
+         workflow, there should be an entry under `meta.outputs` with that same name. Additionally, \
+         these entries should be in the same order (that order is up to the developer to decide). \
+         No extraneous output entries are allowed."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Completeness])
+    }
+}
+
+/// Check each output key exists in the `outputs` key within the `meta` section.
+fn check_output_meta(
+    state: &mut Diagnostics,
+    meta: &MetadataSection,
+    outputs: &OutputSection,
+    context: TaskOrWorkflow,
+) {
+    // Get the `outputs` entries from the `meta` section.
+    if let Some(meta_outputs_key) = meta
+        .items()
+        .find(|entry| entry.name().as_str() == "outputs")
+    {
+        let actual: IndexMap<_, _> = meta_outputs_key
+            .value()
+            .unwrap_object()
+            .items()
+            .map(|entry| {
+                (
+                    entry.name().syntax().to_string(),
+                    entry.syntax().text_range().to_span(),
+                )
+            })
+            .collect();
+
+        // Get the declared outputs.
+        let expected: IndexMap<_, _> = outputs
+            .declarations()
+            .map(|d| {
+                (
+                    d.name().syntax().to_string(),
+                    d.syntax().text_range().to_span(),
+                )
+            })
+            .collect();
+
+        let mut extra_found = false;
+
+        // Check for entries missing from `meta`.
+        for (name, span) in &expected {
+            if !actual.contains_key(name) {
+                if !extra_found {
+                    extra_found = true;
+                }
+                state.add(nonmatching_output(*span, name, &context));
+            }
+        }
+
+        // Check for extra entries in `meta`.
+        for (name, span) in &actual {
+            if !expected.contains_key(name) {
+                if !extra_found {
+                    extra_found = true;
+                }
+                state.add(extra_output_in_meta(*span, name, &context));
+            }
+        }
+
+        // Check for out-of-order entries.
+        if !extra_found && !actual.keys().eq(expected.keys()) {
+            state.add(out_of_order(
+                meta_outputs_key.syntax().text_range().to_span(),
+                outputs.syntax().text_range().to_span(),
+                &context,
+            ));
+        }
+    } else {
+        state.add(missing_outputs_in_meta(
+            meta.syntax().first_token().unwrap().text_range().to_span(),
```

Comment:
Personally, I think the span here should be the full meta section (`span_of(meta)`), but I know we've taken a different approach on other rules so I am fine if you want to keep it this way.